src/main/scala/com/spotify/ratatool/samplers/BigSampler.scala

```diff
@@ -0,0 +1,346 @@
+/*
+ * Copyright 2017 Spotify AB.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.spotify.ratatool.samplers
+
+import java.io.IOException
+import java.net.URI
+import java.nio.charset.Charset
+
+import com.google.api.client.googleapis.services.AbstractGoogleClientRequest
+import com.google.api.client.util.{BackOff, BackOffUtils, Sleeper}
+import com.google.api.services.bigquery.model.{Table, TableFieldSchema, TableReference, TableSchema}
+import com.google.cloud.dataflow.sdk.io.BigQueryIO
+import com.google.cloud.dataflow.sdk.options.BigQueryOptions
+import com.google.cloud.dataflow.sdk.transforms.SerializableFunction
+import com.google.cloud.dataflow.sdk.util.{FluentBackoff, Transport}
+import com.google.cloud.hadoop.util.ApiErrorExtractor
+import com.google.common.hash.{HashCode, HashFunction, Hasher, Hashing}
+import com.google.protobuf.ByteString
+import com.spotify.ratatool.io.{AvroIO, FileStorage}
+import com.spotify.ratatool.serde.JsonSerDe
+import com.spotify.scio.bigquery.TableRow
+import com.spotify.scio.io.{Tap, Taps}
+import com.spotify.scio.{ContextAndArgs, ScioContext, _}
+import org.apache.avro.Schema
+import org.apache.avro.Schema.{Field, Type}
+import org.apache.avro.generic.GenericRecord
+import org.joda.time.Duration
+import org.slf4j.LoggerFactory
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+import scala.concurrent.Future
+import scala.language.existentials
+import scala.util.{Failure, Success, Try}
+
+
+object BigSampler {
+
+  private val log = LoggerFactory.getLogger(BigSampler.getClass)
+  private val utf8Charset = Charset.forName("UTF-8")
+
+  private val kissHashFun = Hashing.murmur3_128(System.currentTimeMillis.toInt)
+  private def kissHashFun(seed: Option[Int]): HashFunction = {
+    seed match {
+      case Some(s) =>  Hashing.murmur3_128(s)
+      case None => kissHashFun
+    }
+  }
+
+  private case class PathsAndSchemas(input: URI,
+                                     output: URI,
+                                     schema: Schema,
+                                     sampleSchema: Class[_])
+
+  // scalastyle:off cyclomatic.complexity
+  private def hashAvroField(r: GenericRecord, fields: Seq[Field], hasher: Hasher): Hasher = {
+    fields.foldLeft(hasher) { case (h, field) =>
+      //TODO: is there need to check for nulls in v?
+      val v = r.get(field.name())
```

Comment:
Yes should probably check `null`s since avro allows them and they're everywhere.